src/Kiota.Builder/KiotaBuilder.cs

```diff
@@ -1548,43 +1548,56 @@ private CodeType CreateModelDeclarationAndType(OpenApiUrlTreeNode currentNode, O
             TypeDefinition = codeDeclaration,
         };
     }
-    private CodeType CreateInheritedModelDeclaration(OpenApiUrlTreeNode currentNode, OpenApiSchema schema, OpenApiOperation? operation, string classNameSuffix, CodeNamespace codeNamespace, bool isRequestBody)
+    private CodeType CreateInheritedModelDeclarationAndType(OpenApiUrlTreeNode currentNode, OpenApiSchema schema, OpenApiOperation? operation, string classNameSuffix, CodeNamespace codeNamespace, bool isRequestBody, string typeNameForInlineSchema)
     {
-        var allOfs = schema.AllOf.FlattenSchemaIfRequired(static x => x.AllOf);
-        CodeElement? codeDeclaration = null;
-        var className = string.Empty;
-        var codeNamespaceFromParent = GetShortestNamespace(codeNamespace, schema);
-        foreach (var currentSchema in allOfs)
+        return new CodeType
         {
-            var referenceId = GetReferenceIdFromOriginalSchema(currentSchema, schema);
-            var shortestNamespaceName = GetModelsNamespaceNameFromReferenceId(referenceId);
-            var shortestNamespace = string.IsNullOrEmpty(referenceId) ? codeNamespaceFromParent : rootNamespace?.FindOrAddNamespace(shortestNamespaceName);
-            className = (currentSchema.GetSchemaName() is string cName && !string.IsNullOrEmpty(cName) ?
+            TypeDefinition = CreateInheritedModelDeclaration(currentNode, schema, operation, classNameSuffix, codeNamespace, isRequestBody, typeNameForInlineSchema),
+        };
+    }
+    private CodeClass CreateInheritedModelDeclaration(OpenApiUrlTreeNode currentNode, OpenApiSchema schema, OpenApiOperation? operation, string classNameSuffix, CodeNamespace codeNamespace, bool isRequestBody, string typeNameForInlineSchema)
+    {
+        var flattenedAllOfs = schema.AllOf.FlattenSchemaIfRequired(static x => x.AllOf).ToArray();
+        var referenceId = schema.Reference?.Id;
+        var className = (schema.GetSchemaName(true) is string cName && !string.IsNullOrEmpty(cName) ?
```

Comment:
```suggestion
        var className = (schema.GetSchemaName() is string cName && !string.IsNullOrEmpty(cName) ?
```
Out of curiosity, is there a specific reason for this change? As we may make an attempt to flatten the schema later on, does it make sense to avoid it when pulling out the name? The `className` is what will make the model unique in the namespace, I believe we may end up with a duplicate class with different names in the same namespace if the schema is flattened due to not having any root information (like a single AllOf/OneOf/AnyOf entry) and end up using the `typeNameForInlineSchema`.

```suggestion
        var className = (schema.GetSchemaName(schema.IsSemanticallyMeaningful()) is string cName && !string.IsNullOrEmpty(cName) ?
```
Alternatively, does it make sense to pass/use a calculated value like `schema.IsSemanticallyMeaningful()` to make the model generation consistent with the model/schema evaluation in the scenarios where we have a root schema that isn't meaningful and is being unwrapped?

Reply:
Thanks for starting a discussion on this.
Why is it set to true: because if it wasn't set, some tests would fail with children taking the name of their parents at this step.
Why is is not using a dynamic value: similarly I had tests failing in the inheritance structure.

We might ultimately have to resolve to a dynamic value + pattern matching at this step in the future. But as of now, all unit, integration and smoke tests are passing. If you can come up with a unit test that fails this step, I'm happy to re-evaluate.